### PR TITLE
Software MFA

### PR DESCRIPTION
**Pull request by DmitryProskurin**
Add methods for software MFA setup via authorized Cognito user.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by DmitryProskurin** [comment]
@normj, could you review my PR?

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs [hunk 1]
```diff
@@ -554,6 +554,62 @@ public virtual async Task<ListDevicesResponse> ListDevicesV2Async(int limit, str
             return await Provider.ListDevicesAsync(listDevicesRequest, cancellationToken).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Request code for authenticator app.
+        /// </summary>
+        /// <returns><see cref="AssociateSoftwareTokenResponse"/> with secret code.</returns>
+        public virtual async Task<AssociateSoftwareTokenResponse> AssociateSoftwareTokenAsync()
```
Comment (peterrsongg) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs:561]: Can you add versions with `cancellationToken` for these methods?

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs [hunk 2]
```diff
@@ -554,6 +554,62 @@ public virtual async Task<ListDevicesResponse> ListDevicesV2Async(int limit, str
...
+            return await Provider.AssociateSoftwareTokenAsync(request).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Verify code from authenticator app.
+        /// </summary>
+        /// <param name="code"></param>
+        /// <returns><see cref="VerifySoftwareTokenResponse"/> which contains token verification status.</returns>
+        public virtual async Task<VerifySoftwareTokenResponse> VerifySoftwareTokenAsync(string code) {
+            if (string.IsNullOrEmpty(code))
```
Comment (peterrsongg) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs:579]: Same comment as above

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs [hunk 3]
```diff
@@ -554,6 +554,62 @@ public virtual async Task<ListDevicesResponse> ListDevicesV2Async(int limit, str
...
+        }
+
+        /// <summary>
+        /// Update settings for software MFA settings.
+        /// </summary>
+        /// <param name="isPreferred">Software MFA preferred at sign in.</param>
+        /// <param name="isEnabled">Enable or disable software MFA.</param>
+        /// <returns></returns>
+        public async Task UpdateSoftwareMfaSettingsAsync(bool isPreferred, bool isEnabled)
+        {
```
Comment (peterrsongg) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs:600]: Same comment, add a version of the method with `cancellationToken`

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs (hunk 1 above)
Comment (DmitryProskurin) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs]: Done.

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs (hunk 2 above)
Comment (DmitryProskurin) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs:581]: Done.

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs (hunk 3 above)
Comment (DmitryProskurin) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs:603]: Done.

**Review by normj — APPROVED** [review]

**Review by peterrsongg — APPROVED** [review]

**Comment by normj** [comment]
Thanks for the PR! It has been released as part of version 2.5.0.

**Pull request merged**